File: rust_stub/src/controller_out/x360.rs
use crate::{
    controller_abs::{
        Axis, BitPackedButton, BitPackedButtons, Gamepad, GamepadAxis, GamepadButton, JoystickState,
    },
    AxisNew, JoystickStateNew,
};
use std::{u8, vec};

pub struct XboxButtonState {
    pub a: BitPackedButton,
    pub b: BitPackedButton,
    pub x: BitPackedButton,
    pub y: BitPackedButton,
    pub lb: BitPackedButton,
    pub rb: BitPackedButton,
    pub l3: BitPackedButton,
    pub r3: BitPackedButton,
    pub start: BitPackedButton,
    pub options: BitPackedButton,
    pub dpad_up: BitPackedButton,
    pub dpad_down: BitPackedButton,
    pub dpad_left: BitPackedButton,
    pub dpad_right: BitPackedButton,
    pub xbox: BitPackedButton,
}

impl XboxButtonState {
    pub fn new() -> XboxButtonState {
        XboxButtonState {
            a: BitPackedButton::new("A".to_string(), 0x04),
            b: BitPackedButton::new("B".to_string(), 0x05),
            x: BitPackedButton::new("X".to_string(), 0x06),
            y: BitPackedButton::new("Y".to_string(), 0x07),
            lb: BitPackedButton::new("LB".to_string(), 0x00),
            rb: BitPackedButton::new("RB".to_string(), 0x01),
            // Joystick buttons
            l3: BitPackedButton::new("L3".to_string(), 0x06),
            r3: BitPackedButton::new("R3".to_string(), 0x07),

            start: BitPackedButton::new("START".to_string(), 0x04),
            options: BitPackedButton::new("OPTIONS".to_string(), 0x05),
            xbox: BitPackedButton::new("XBOX".to_string(), 0x02),

            // Dpad
            dpad_up: BitPackedButton::new("DPAD_UP".to_string(), 0x00),
            dpad_down: BitPackedButton::new("DPAD_DOWN".to_string(), 0x01),
            dpad_left: BitPackedButton::new("DPAD_LEFT".to_string(), 0x02),
            dpad_right: BitPackedButton::new("DPAD_RIGHT".to_string(), 0x03),
        }
    }

    pub fn get_control_byte_2(&self) -> u8 {
        BitPackedButtons {
            buttons: vec![
                self.dpad_up.clone(),
                self.dpad_down.clone(),
                self.dpad_left.clone(),
                self.dpad_right.clone(),
                self.start.clone(),
                self.options.clone(),
                self.l3.clone(),
                self.r3.clone(),
            ],
        }
        .to_bytes_repr()
    }

    pub fn get_control_byte_3(&self) -> u8 {
        BitPackedButtons {
            buttons: vec![
                self.lb.clone(),
                self.rb.clone(),
                self.xbox.clone(),
                self.a.clone(),
                self.b.clone(),
                self.x.clone(),
                self.y.clone(),
            ],
        }
        .to_bytes_repr()
    }
}

pub struct XboxControllerState {
    pub buttons: XboxButtonState,
    pub left_trigger: Axis,
    pub right_trigger: Axis,
    pub left_joystick: JoystickState,  // byte 6 - 9
    pub right_joystick: JoystickState, // byte 10 - 13
}

impl XboxControllerState {
    pub fn new() -> XboxControllerState {
        XboxControllerState {
            buttons: XboxButtonState::new(),
            left_trigger: AxisNew!(u8::MIN),
            right_trigger: AxisNew!(u8::MIN),
            left_joystick: JoystickStateNew!(i16, 0),
            right_joystick: JoystickStateNew!(i16, 0),
        }
    }

    pub fn update_from_gamepad(&mut self, gamepad: &Gamepad) {
        for (button, button_state) in &gamepad.buttons {
            let val = button_state.to_owned();
            match button {
                GamepadButton::North => self.buttons.y.value = val,
                GamepadButton::East => self.buttons.b.value = val,
                GamepadButton::South => self.buttons.a.value = val,
                GamepadButton::West => self.buttons.x.value = val,
                GamepadButton::LeftShoulderButton => self.buttons.lb.value = val,
                GamepadButton::RightShoulderButton => self.buttons.rb.value = val,
                GamepadButton::LeftThumb => self.buttons.l3.value = val,
                GamepadButton::RightThumb => self.buttons.r3.value = val,
                GamepadButton::Start => self.buttons.start.value = val,
                GamepadButton::Select => self.buttons.options.value = val,
                GamepadButton::Mode => self.buttons.xbox.value = val,
                GamepadButton::DPadUp => self.buttons.dpad_up.value = val,
                GamepadButton::DPadDown => self.buttons.dpad_down.value = val,
                GamepadButton::DPadLeft => self.buttons.dpad_left.value = val,
                GamepadButton::DPadRight => self.buttons.dpad_right.value = val,
            }
        }
        for (gamepad_axis, axis) in &gamepad.axes {
            match gamepad_axis {
                GamepadAxis::LeftJoystickX => {
                    self.left_joystick.x.value = axis.convert_into(false);
                }
                GamepadAxis::LeftJoystickY => {
                    self.left_joystick.y.value = axis.convert_into(false);
                }
                GamepadAxis::RightJoystickX => {
                    self.right_joystick.x.value = axis.convert_into(false);
                }
                GamepadAxis::RightJoystickY => {
                    self.right_joystick.y.value = axis.convert_into(false);
                }
                GamepadAxis::LeftTrigger => {
                    self.left_trigger.value = axis.convert_into(false);
                }
                GamepadAxis::RightTrigger => {
                    self.right_trigger.value = axis.convert_into(false);
                }
            }
        }
    }

    pub fn to_packet(&self) -> [u8; 20] {
        let mut packet = [0u8; 20];
        packet[0] = 0x00; // Report ID (0x00)
        packet[1] = 0x14; // Length (0x14)
        packet[2] = self.buttons.get_control_byte_2();
        packet[3] = self.buttons.get_control_byte_3();
        packet[4] = self.left_trigger.convert_into(false);
        packet[5] = self.right_trigger.convert_into(false);
        packet[6..8].copy_from_slice(
            &self
                .left_joystick
                .x
                .convert_into::<i16, _>(false)
                .to_le_bytes(),
        );
        packet[8..10].copy_from_slice(
            &self
                .left_joystick
                .y
                .convert_into::<i16, _>(false)
                .to_le_bytes(),
        );
        packet[10..12].copy_from_slice(
            &self
                .right_joystick
                .x
                .convert_into::<i16, _>(false)
                .to_le_bytes(),
        );
        packet[12..14].copy_from_slice(
            &self
                .right_joystick
                .y
                .convert_into::<i16, _>(false)
                .to_le_bytes(),
        );
        packet
    }
}

File: python/x360.py
from dataclasses import dataclass
import struct
from enum import Enum
from evdev import ecodes


class X360Surfaces(Enum):
    A = ("A", ecodes.BTN_A)
    B = ("B", ecodes.BTN_B)
    X = ("X", ecodes.BTN_X)
    Y = ("Y", ecodes.BTN_Y)
    OPTIONS = ("OPTIONS", ecodes.BTN_SELECT)
    XBOX = ("XBOX", ecodes.BTN_MODE)
    START = ("START", ecodes.BTN_START)
    STICK_L = ("STICK_L", ecodes.BTN_THUMBL)
    STICK_R = ("STICK_R", ecodes.BTN_THUMBR)
    BUMPER_L = (("BUMPER_L", ecodes.BTN_TL),)
    BUMPER_R = ("BUMPER_R", ecodes.BTN_TR)

    # Dpad
    DPAD_UP = ("DPAD_UP", ecodes.ABS_HAT0Y)  # -1
    DPAD_DOWN = ("DPAD_DOWN", ecodes.ABS_HAT0Y)  # 1
    DPAD_LEFT = ("DPAD_LEFT", ecodes.ABS_HAT0X)  # -1
    DPAD_RIGHT = ("DPAD_RIGHT", ecodes.ABS_HAT0X)  # 1

    # Triggers
    LEFT_TRIGGER = ("LEFT_TRIGGER", ecodes.ABS_BRAKE)  # 8-bit 0-1024
    RIGHT_TRIGGER = ("RIGHT_TRIGGER", ecodes.ABS_GAS)  # 8-bit 0-1024

    # Joystick axis
    LEFT_JOYSTICK_X = ("LEFT_JOYSTICK_X", ecodes.ABS_X)
    LEFT_JOYSTICK_Y = ("LEFT_JOYSTICK_Y", ecodes.ABS_Y)
    RIGHT_JOYSTICK_X = ("RIGHT_JOYSTICK_X", ecodes.ABS_RZ)
    RIGHT_JOYSTICK_Y = ("RIGHT_JOYSTICK_Y", ecodes.ABS_Z)


@dataclass
class BitPackedButton:
    name: str
    bit: int
    value: bool = False


class BitPackedButtons:
    def __init__(self, buttons):
        self.buttons = buttons

    def to_bytes_repr(self) -> int:
        result = 0
        for b in self.buttons:
            if b.value:
                result |= 1 << b.bit
        return result & 0xFF


@dataclass
class Axis:
    value: int = 0  # default
    bits: int = 8  # triggers are 8-bit by default

    def to_u8(self) -> int:
        return max(0, min(255, self.value))


@dataclass
class JoystickAxis:
    value: int = 0

    def to_i16(self) -> int:
        return max(-32768, min(32767, self.value))


@dataclass
class JoystickState:
    x: JoystickAxis
    y: JoystickAxis


class XboxButtonState:
    def __init__(self):
        self.a = BitPackedButton("A", 0x04)
        self.b = BitPackedButton("B", 0x05)
        self.x = BitPackedButton("X", 0x06)
        self.y = BitPackedButton("Y", 0x07)
        self.lb = BitPackedButton("LB", 0x00)
        self.rb = BitPackedButton("RB", 0x01)
        self.l3 = BitPackedButton("L3", 0x06)
        self.r3 = BitPackedButton("R3", 0x07)
        self.start = BitPackedButton("START", 0x04)
        self.options = BitPackedButton("OPTIONS", 0x05)
        self.xbox = BitPackedButton("XBOX", 0x02)
        self.dpad_up = BitPackedButton("DPAD_UP", 0x00)
        self.dpad_down = BitPackedButton("DPAD_DOWN", 0x01)
        self.dpad_left = BitPackedButton("DPAD_LEFT", 0x02)
        self.dpad_right = BitPackedButton("DPAD_RIGHT", 0x03)

        self.all = [
            self.a,
            self.b,
            self.x,
            self.y,
            self.lb,
            self.rb,
            self.l3,
            self.r3,
            self.start,
            self.options,
            self.xbox,
            self.dpad_up,
            self.dpad_down,
            self.dpad_left,
            self.dpad_right,
        ]

    def get_control_byte_2(self) -> int:
        return BitPackedButtons(
            [
                self.dpad_up,
                self.dpad_down,
                self.dpad_left,
                self.dpad_right,
                self.start,
                self.options,
                self.l3,
                self.r3,
            ]
        ).to_bytes_repr()

    def get_control_byte_3(self) -> int:
        return BitPackedButtons(
            [
                self.lb,
                self.rb,
                self.xbox,
                self.a,
                self.b,
                self.x,
                self.y,
            ]
        ).to_bytes_repr()

    def get_button(self, name: str):
        """Return a BitPackedButton by its name (case-insensitive)."""
        name = name.upper()
        for btn in self.all:
            if btn.name.upper() == name:
                return btn
        raise ValueError(f"No button found with name '{name}'")


class XboxControllerState:
    def __init__(self):
        self.buttons = XboxButtonState()
        self.left_trigger = Axis(0, 8)
        self.right_trigger = Axis(0, 8)
        self.left_joystick = JoystickState(JoystickAxis(0), JoystickAxis(0))
        self.right_joystick = JoystickState(JoystickAxis(0), JoystickAxis(0))

    def to_packet(self) -> bytes:
        packet = bytearray(20)
        packet[0] = 0x00  # Report ID
        packet[1] = 0x14  # Length
        packet[2] = self.buttons.get_control_byte_2()
        packet[3] = self.buttons.get_control_byte_3()
        packet[4] = self.left_trigger.to_u8()
        packet[5] = self.right_trigger.to_u8()

        # pack little-endian signed 16-bit joystick values
        packet[6:8] = struct.pack("<h", self.left_joystick.x.to_i16())
        packet[8:10] = struct.pack("<h", self.left_joystick.y.to_i16())
        packet[10:12] = struct.pack("<h", self.right_joystick.x.to_i16())
        packet[12:14] = struct.pack("<h", self.right_joystick.y.to_i16())

        return bytes(packet)

    def by_enum(self, enum_member: X360Surfaces):
        """Return the corresponding controller attribute for the given X360Surfaces enum."""
        mapping = {
            X360Surfaces.A: self.buttons.a,
            X360Surfaces.B: self.buttons.b,
            X360Surfaces.X: self.buttons.x,
            X360Surfaces.Y: self.buttons.y,
            X360Surfaces.BUMPER_L: self.buttons.lb,
            X360Surfaces.BUMPER_R: self.buttons.rb,
            X360Surfaces.STICK_L: self.buttons.l3,
            X360Surfaces.STICK_R: self.buttons.r3,
            X360Surfaces.START: self.buttons.start,
            X360Surfaces.OPTIONS: self.buttons.options,
            X360Surfaces.XBOX: self.buttons.xbox,
            X360Surfaces.DPAD_UP: self.buttons.dpad_up,
            X360Surfaces.DPAD_DOWN: self.buttons.dpad_down,
            X360Surfaces.DPAD_LEFT: self.buttons.dpad_left,
            X360Surfaces.DPAD_RIGHT: self.buttons.dpad_right,
            X360Surfaces.LEFT_TRIGGER: self.left_trigger,
            X360Surfaces.RIGHT_TRIGGER: self.right_trigger,
            X360Surfaces.LEFT_JOYSTICK_X: self.left_joystick.x,
            X360Surfaces.LEFT_JOYSTICK_Y: self.left_joystick.y,
            X360Surfaces.RIGHT_JOYSTICK_X: self.right_joystick.x,
            X360Surfaces.RIGHT_JOYSTICK_Y: self.right_joystick.y,
        }
        try:
            return mapping[enum_member]
        except KeyError:
            raise ValueError(f"No control found for enum {enum_member}")
